### PR TITLE
Animation sheet golden test

### PR DESCRIPTION
**Pull request by dkwingsmt**
## Description

This PR adds support for golden-testing animating widgets in an animation sheet.

The change introduces 2 new APIs:
* `WidgetTester.pumpFrames`, which is like `pump` but pumps a number of frames with the specified interval of `frameDuration` 
* `SheetDisplay`, which places clones of the target widget in a grid, and assign them with keys from the tail to the front

In this way, clones of the widget will be added to the front of the sheet one every frame, so that after a number of frames, the sheet is filled with clones that show each frame of the widget.

Take `CircularProgressionIndicator` as an example. Tests such as follows:

```dart
  testWidgets('Indeterminate CircularProgressIndicator uses expected animation', (WidgetTester tester) async {
    await tester.pumpFrames(
      (int frameIndex) => SheetDisplay(
        cellNum: frameIndex + 1,
        cellSize: const Size(48, 48),
        target: const Padding(
          padding: EdgeInsets.all(4.0),
          child: CircularProgressIndicator(),
        ),
      ),
      16 * 12,
    );

    expect(
      find.byType(SheetDisplay),
      matchesGoldenFile('circular_progress_indicator.grid.png'),
    );
  });
```

Will yield a result as:

circular_progress_indicator grid

## Related Issues

- Blocks https://github.com/flutter/flutter/pull/50412, which needs an animation test
- Affected by https://github.com/flutter/flutter/issues/54511, golden tests yield incorrect result on chrome

## Tests

I added the following tests:

- `pumpFrames`

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_test/lib/src/sheet_display.dart [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  @override
+  void paint(PaintingContext context, Offset offset) {
+    defaultPaint(context, offset);
+    return;
+  }
+
+  @override
+  bool get sizedByParent => true;
```
Comment (Hixie) [inline packages/flutter_test/lib/src/sheet_display.dart:152]: we usually put this above performResize/performLayout

File: packages/flutter_test/lib/src/sheet_display.dart [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    final int maxChildrenNum = _columnNum * _rowNum;
+    int childrenNum = 0;
+    RenderBox child = firstChild;
+    double x = 0;
+    double y = 0;
+    while (child != null && childrenNum < maxChildrenNum) {
+      assert(y + cellSize.height < size.height);
+      assert(x + cellSize.width < size.width);
+      final _CellGridParentData childParentData = child.parentData as _CellGridParentData;
+      child.layout(BoxConstraints.tight(cellSize), parentUsesSize: false);
```
Comment (Hixie) [inline packages/flutter_test/lib/src/sheet_display.dart:133]: performResize must not call child.layout

Comment (Hixie) [reply packages/flutter_test/lib/src/sheet_display.dart:133]: i'm surprised that doesn't fire an assert. we should add one. cc @goderbauer

File: packages/flutter_test/lib/src/sheet_display.dart [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+  @override
+  void setupParentData(RenderBox child) {
+    if (child.parentData is! _CellGridParentData)
+      child.parentData = _CellGridParentData();
+  }
+
+  @override
+  void performResize() {
+    size = constraints.biggest;
+    final int maxChildrenNum = _columnNum * _rowNum;
```
Comment (Hixie) [inline packages/flutter_test/lib/src/sheet_display.dart:124]: basically line 123 should be the body of performResize and the rest should be in performLayout

File: packages/flutter_test/lib/src/widget_tester.dart [hunk 4]
```diff
@@ -545,6 +548,28 @@ class WidgetTester extends WidgetController implements HitTestDispatcher, Ticker
     }).then<int>((_) => count);
   }
 
+  /// Repeatedly triggers `frameCount` frames with a fixed `frameDuration`.
+  /// 
+  /// The `frameDuration` defaults to 16.683 milliseconds (59.94 FPS).
+  /// All frames stop executing after finishing [EnginePhase.build], except for
+  /// the last frame, which executes the full pass.
```
Comment (Hixie) [inline packages/flutter_test/lib/src/widget_tester.dart:555]: Why the weird behaviour with the EnginePhase? Seems like it's asking for trouble (e.g. what if you have a LayoutBuilder?)

Comment (dkwingsmt) [reply packages/flutter_test/lib/src/widget_tester.dart:555]: I'm concerned about the executing time, which can be dropped from 1.3s to 0.7s in this way (while all other tests, including normal golden tests). Should I remove it?

File: packages/flutter_test/lib/src/sheet_display.dart [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    @required this.cellSize,
+    @required this.cellNum,
+  }) : assert(cellSize != null),
+       assert(cellNum != null),
+       super(key: key);
+
+  /// The widget to show animation of.
+  final Widget target;
+
+  /// The size of a cell. Should contain a widget.
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/sheet_display.dart:35]: What do you mean by "Should contain a widget"? A `Size` can't have a `Widget`?

File: packages/flutter_test/lib/src/sheet_display.dart (hunk 2 above)
Comment (goderbauer) [reply packages/flutter_test/lib/src/sheet_display.dart:133]: Filed https://github.com/flutter/flutter/issues/54767.

File: packages/flutter_test/lib/src/sheet_display.dart [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+  /// 
+  /// Should not change throughout the test.
+  final Size cellSize;
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/sheet_display.dart:38]: What happens if `target` wants to be smaller then `cellSize`?

File: packages/flutter_test/lib/src/sheet_display.dart [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+  }) : assert(cellSize != null),
+       assert(cellNum != null),
+       super(key: key);
+
+  /// The widget to show animation of.
+  final Widget target;
+
+  /// The size of a cell. Should contain a widget.
+  /// 
+  /// Should not change throughout the test.
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/sheet_display.dart:37]: Why? The widget would just re-layout with the changed size, no?

File: packages/flutter_test/lib/src/sheet_display.dart [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+import 'package:flutter/rendering.dart';
+import 'package:flutter/widgets.dart';
+
+/// Place `cellNum` number of the target widget in a grid sheet.
+/// 
+/// The [SheetDisplay] tries to fill as much space as the parent allows, then
+/// display clones of the target widget in cells of size `cellSize`, placing
+/// cells from top left to bottom right, horizontal first. Cells are keyed from
+/// the tail to the front, meaning new clones are added to the front of the list.
+///
```
Comment (Piinks) [inline packages/flutter_test/lib/src/sheet_display.dart:14]: Can you include a sample?

File: packages/flutter_test/lib/src/widget_tester.dart [hunk 9]
```diff
@@ -545,6 +548,28 @@ class WidgetTester extends WidgetController implements HitTestDispatcher, Ticker
...
+  Future<void> pumpFrames(
```
Comment (Piinks) [inline packages/flutter_test/lib/src/widget_tester.dart:556]: Could the docs also provide a sample invocation? I wasn't quite sure how to use it until I looked at #54335

**Review by Piinks — COMMENTED** [review]
Is it possible to interact with the target widget? Some of our animations are triggered by a gesture, like https://github.com/flutter/flutter/blob/6a983e734151dbd6ddcc7457bb3ce24a7212f331/packages/flutter/test/material/bottom_navigation_bar_test.dart#L1443 which is a test that would hope to use this feature.The navigation bar is tapped and the captured frames validate the ink splash animation.

Also, I am not sure how to use this properly. I've tried a couple of different animations but have not been successful. Maybe some documentation would help? In particular, how do you determine the correct cell size? I tried to put an existing golden file in a SheetDisplay, and gave it the size of the golden file, but it yielded several exceptions.

**Comment by goderbauer** [comment]
(PR triage fly-by): Cirrus is unhappy.

File: packages/flutter_test/lib/src/widget_tester.dart (hunk 4 above)
Comment (dkwingsmt) [reply packages/flutter_test/lib/src/widget_tester.dart:555]: I've removed this behavior. Now all frames are pumped through.

File: packages/flutter_test/lib/src/sheet_display.dart (hunk 7 above)
Comment (dkwingsmt) [reply packages/flutter_test/lib/src/sheet_display.dart:37]: I've removed this requirement.

File: packages/flutter_test/lib/src/sheet_display.dart (hunk 6 above)
Comment (dkwingsmt) [reply packages/flutter_test/lib/src/sheet_display.dart:38]: If the test target wants to be smaller than `cellSize` then it should be wrapped in a `Center` or something. Or just provide a proper `cellSize`.

Comment (goderbauer) [reply packages/flutter_test/lib/src/sheet_display.dart:38]: Maybe document that in the comment above.

**Comment by dkwingsmt** [comment]
Close for now until the major changes have been finished

**Pull request closed without merging**